osdconfig/interface.go

```diff
@@ -0,0 +1,63 @@
+// osdconfig is a package to work with distributed config parameters
+package osdconfig
+
+import "context"
+
+// A config manager interface allows management of osdconfig parameters
+// It defines setters, getters and callback management functions
+type ConfigManager interface {
+	// GetClusterConf fetches cluster configuration data from a backend such as kvdb
+	GetClusterConf() (*ClusterConfig, error)
+
+	// Fetch node configuration data using node id
+	GetNodeConf(nodeID string) (*NodeConfig, error)
+
+	// SetClusterConf pushes cluster configuration data to the backend
+	// It is assumed that the backend will notify the implementor of this interface
+	// when a change is triggered
+	SetClusterConf(config *ClusterConfig) error
+
+	// SetNodeConf pushes node configuration data to the backend
+	// It is assumed that the backend will notify the implementor of this interface
+	// when a change is triggered
+	SetNodeConf(config *NodeConfig) error
+
+	// Register callback functions that will get triggered on a change to either
+	// cluster configuration or node configuration data in the backend.
+	// The callback function returns two channels: A channel to write to and a
+	// channel to read on. Context is provided to the callback function in order
+	// to manage cleanup during context cancellations
+	Register(name string, watcherType Watcher,
+		opt interface{},
+		cb func(ctx context.Context, opt interface{}) (chan<- *DataToCallback,
```

Comment:
Maybe clearer to name them readChannel and writeChannel?

Reply:
i like that. I used send/recv elsewhere in this code but read/write is much nicer.

Reply:
done